lib/appoptics_apm/inst/grpc_client.rb

```diff
@@ -0,0 +1,161 @@
+        AppOpticsAPM::SDK.set_transaction_name(metadata[:method]) if AppOpticsAPM.transaction_name.nil?
+
+        patch_set_input_stream_done
+
+        response = bidi_streamer_without_appoptics(req, metadata: metadata)
+        block_given? ? response.each { |r| yield r } : response
+      rescue => e
+        unless e.instance_variable_get(:@exn_logged)
+          context_from_incoming
+          AppOpticsAPM::API.log_exception('grpc_client', e)
+          AppOpticsAPM::API.log_exit('grpc_client', exit_tags(@tags))
+        end
+        raise e
+      end
+
+      private
+
+      def unary_response(req, type: , metadata: , without:)
+        tags = grpc_tags(type, metadata[:method] || metadata_to_send[:method])
+        AppOpticsAPM::SDK.trace('grpc_client', tags) do
+          metadata['x-trace'] = AppOpticsAPM::Context.toString
+          AppOpticsAPM::SDK.set_transaction_name(metadata[:method]) if AppOpticsAPM.transaction_name.nil?
+          begin
+            send(without, req, metadata: metadata)
+          ensure
+            exit_tags(tags)
+            context_from_incoming
+          end
+        end
+      end
+
+      def patch_receive_and_check_status
+        def self.receive_and_check_status # need to patch this so that log_exit can be called after the enum is consumed
+          super
+          context_from_incoming
+        rescue => e
+          context_from_incoming
+          AppOpticsAPM::API.log_exception('grpc_client', e)
+          raise e
+        ensure
+          AppOpticsAPM::API.log_exit('grpc_client', exit_tags(@tags))
+        end
+      end
+
+      def patch_set_input_stream_done
+        # need to patch this instance method so that log_exit can be called after the enum is consumed
+        def self.set_input_stream_done
+          return if status.nil?
+          context_from_incoming
+          if status.code > 0
+            AppOpticsAPM::API.log_exception('grpc_client', $!)
+          end
+          AppOpticsAPM::API.log_exit('grpc_client', exit_tags(@tags))
+          super
+        end
+      end
+
+      def context_from_incoming
+        xtrace ||= @call.trailing_metadata['x-trace'] if @call.trailing_metadata && @call.trailing_metadata['x-trace']
+        xtrace ||= @call.metadata['x-trace'] if @call.metadata && @call.metadata['x-trace']
+        AppOpticsAPM::Context.fromString(xtrace) if xtrace
+      end
+
+      def exit_tags(tags)
+        # we need to translate the status.code, it is not the status.details we want, they are not matching 1:1
+        tags['GRPCStatus'] ||= @call.status ? StatusCodes[@call.status.code].to_s : 'UNKNOWN'
+        tags.delete('Backtrace')
+        tags
+      end
+    end
+
+  end
+end
+
+if defined?(::GRPC) && AppOpticsAPM::Config[:grpc_client][:enabled]
+  AppOpticsAPM.logger.info '[appoptics_apm/loading] Instrumenting GRPC' if AppOpticsAPM::Config[:verbose]
+
+  # Client side is instrumented in ActiveCall and ClientStub
```

Comment:
This comment seems odd to me - `ActiveCall` is a module/scope while `ClientStub` is a class. I went looking for the `ActiveCall` class and the `ClientStub` scope before I concluded that they don't exist.

Reply:
`AppOptics::GRPC::ActiveCall` is a module that we are including it in the class `::GRPC::ActiveCall` where the methods from `AppOptics::GRPC::ActiveCall` will become instance methods. 
Good write-up here: https://medium.com/@leo_hetsch/ruby-modules-include-vs-prepend-vs-extend-f09837a5b073
I would love to use `prepend`, because then we could avoid this aliasing stuff:
`::AppOpticsAPM::Util.method_alias(klass, :request_response, ::GRPC::ActiveCall)`. It only became available in Ruby 2xx and I need to test some scenarios where there is other instrumentation and analyze the possibility of interference.
The class `GRPC::ClientStub` is instrumented using a different pattern. Because all 4 methods just need a single additional line `metadata[:method] = method` these methods are defined in the class directly and then there is again an alias_chain created with ::AppOpticsAPM::Util.method_alias(::GRPC::ClientStub, m).

Reply:
thanks. i didn't realize that include was literally bringing the text into the class; i was thinking of it as more of a javascript require where it inserts its own namespace.

so a class can contain methods in multiple scopes? i.e., a class is not just a single scope?

i don't understand alias but i did understand (i think) that you were calling the original, unmodified method, with `send`. this was an example of meta-programming, i.e., writing methods on the fly?